Fixing critical issues with Fees > Batch Payments

zbig01 opened the pull request:
Some code changes in the modernization of the UI rendered parts of
the Batch payemnts pages unusable/difficult to use

Changes are made to the following files:

interface/billing/edit_payment.php
interface/billing/new_payment.php
interface/billing/payment_master.inc.php
interface/billing/payment_pat_sel.inc.php
interface/billing/search_payments.php

with these changes the Batch payments section is once again functional

The issues fixed are:

1. html code populating Payment Method, Paying Entity and Payment Category select boxes when Fees > Batch Payments > New Payment 'Save Changes' button is clicked after filling in valid data

![batch_payment_022](https://user-images.githubusercontent.com/12654295/64508264-dbb9f980-d291-11e9-8102-96173b38f8dd.png)

**Cause:** In interface/billing/payment_master.inc.php which is a part of interface/billing/new_patient.php i.e. Fees > Batch Payments > New Payment form - `function generate_print_field` was being used incorrectly to generate the option values for these three select boxes when the interface/billing/new_patient.php was saved after data entry. This function is designed to write a un-editable table that would display the particulars of the saved batch payment to help in distributing the batch payment and is not designed to generate a select box with the correctly selected option value.

**Solution:** Replaced these three Select boxes with disabled input boxes upon saving the Fees > Batch Payments > New Payment form. These 3 input boxes get their values from `function getListItemTitle` instead. The saved Fees > Batch Payments > New Payment form that is returned after a successful post displays this data as an un-editable form thereby mimicking the original intended functionality.  

2.  Upon clicking the Fees > Batch Payments > New Payment  'Save Changes' button once, the 'Do you want to save' alert box was being triggered twice thus saving same batch payment data twice,      especially if the OK is pressed each time to close the alert box.

    **Cause:** function SavePayment() was being called twice  when  Fees > Batch Payments > New Payment  'Save Changes' button was clicked once. 
        Once in interface/billing/payment_master.inc.php on line 261 - `<button  id="save_btn" onClick="javascript:return SavePayment();" class="btn btn-default btn-save"><?php echo xlt('Save Changes');?></button>`
        The second time was on line 452 in interface/billing/new_patient.php - `echo 'top.restoreSession();return SavePayment();';`
    
    **Solution:** Commented out `echo 'top.restoreSession();return SavePayment();';` on line 452 in  interface/billing/new_patient.php
    
3. Edit Changes popup modal not working when a batch payment is clicked from a filtered list displayed when Fees > Batch Payments > Search Payment > 'Search' button is clicked

    **Cause:** `function loadiframe` was inadvertently removed when switching to `class OemrUI `in the UI modernizing project.
    
    **Solution:** Reinstated `function loadiframe` in interface/billing/search_payments.php. Therefore when Fees > Batch Payments > Search Payment 'Search' button is clicked  a table containing the batch payment results matching filtered by search criteria will be displayed. Upon clicking any particular batch payment a popup modal, Edit Payment displays the batch payment detail along with the undistributed amount. To ensure no clash with the modal that displays the help file the data-target was changed to `"#myModal1"` from `"#myModal"` and the modal's id was changed to `"targetiframe1"`  from `"targetiframe"`
    
`
 function loadiframe(htmlHref) { //load iframe
         document.getElementById('targetiframe1').src = htmlHref;
    }
`

4. Edit Payment popup shows only the batch payment details but not the table with details of the already made payments

    **Cause:** The div containing the table data was floating to the right and having zero width and thus was not visible
    
    **Solution:** enclosed the table in a div with Bootstrap 3 class `<div class="col-xs-12"><div>`
    
    
5. Clicking 'Cancel changes' button in Edit payment modal (interface/billing/edit_payment.php) gives JS error - `Uncaught ReferenceError: ResetForm is not defined at HTMLButtonElement.onclick (edit_payment.php?payment_id=xxxx:yyyy)`

    **Cause:** JS `function ResetForm()` exists only in interface/billing/new_patient.php and thus accessible only when Fees > Batch Payments > New Payment is being accessed.
    
    **Solution:** Copied the `function ResetForm()` to interface/billing/edit_payment.php

stephenwaite commented:
wonderful pull request, :100: 
might get a chance to learn something thanks to your detailed explanation, will pull your branch and test and will be honored to merge into master, thank you @zbig01

bradymiller commented:
Agree with @stephenwaite ! After @stephenwaite brings this in, I'll get it into the first 5.0.2 patch, which should be getting released by this weekend.

stephenwaite commented on interface/billing/payment_master.inc.php:
```diff
@@ -285,10 +285,11 @@ function generate_list_payment_category($tag_name, $list_id, $currvalue, $title,
-                            <input type="text" class="form-control" name="payment_method1" id="payment_method1" value="<?php $frow['data_type']=1;
-                            $frow['list_id']='payment_method';
-                            generate_print_field($frow, $PaymentMethod);?>" disabled />
+                            <label class="control-label" for="payment_method"><?php echo xlt('Payment Method'); ?>:</label>
```
nice catch, probably meant `generate_form_field` and it somehow worked for a long time :)

stephenwaite commented on interface/billing/new_payment.php:
```diff
@@ -449,7 +449,7 @@ function FillUnappliedAmount()
                 if ($payment_id*1==0) {
-                    echo 'top.restoreSession();return SavePayment();';
+                   // echo 'top.restoreSession();return SavePayment();';
                 } else {
```
also remember @bradymiller suggests removing the offending code vs commenting

zbig01 replied:
It did not appear so, but I can reinstate it and remove the return SavePayment statement instead

zbig01 replied:
Retained restoreSession, removed return SavePayment

stephenwaite commented on interface/billing/new_payment.php:
```diff
@@ -509,5 +512,11 @@ function FillUnappliedAmount()
 });
 </script>
+<script>
+//$('#save_btn').on('click', SavePayment);
```
could be removed?

zbig01 replied:
yes

zbig01 replied:
will do

zbig01 replied:
Done

stephenwaite commented on interface/billing/new_payment.php:
```diff
@@ -509,5 +512,11 @@ function FillUnappliedAmount()
 </script>
+<script>
+//$('#save_btn').on('click', SavePayment);
+$('#save_btn').on('click', function(){
```
seeing a btn-save class but no save_btn id?

zbig01 replied:
I was trying to see if the JS could be moved to new_payment. In the end I decided against it, will remove :)

zbig01 replied:
Done

stephenwaite commented on interface/billing/search_payments.php:
```diff
@@ -615,18 +615,18 @@ function SearchPayingEntityAction()
                                 <!--<a class='iframe medium_modal' href="edit_payment.php?payment_id=<?php echo attr_url($RowSearch['session_id']); ?>"><?php
                                                         $frow['data_type']=1;
                                                         $frow['list_id']='payment_method';
                                                         generate_print_field($frow, $RowSearch['payment_method']);
```
looks familiar :), wondering if function is limited like your fix above

zbig01 replied:
It is exactly the same, except to data-target is different.  I did it to not clash with the other modal. Since this was a one off instance I let it be. In case this problem is recurrent the oeUI class will need to be modified

zbig01 commented:
@stephenwaite @bradymiller 

Made the requested changes. Our biller suggested 2 minor changes to billing report that I have incorporated
modified billing_report.php to display highlighted provider name
modified billrep.inc to display list sorted first by provider_id

stephenwaite commented:
thanks for the fixes @zbig01 , can get the minor travis error after bring it in
